File: id-card-toolkit-ios-sdk-v3.1.6/id-card-toolkit-ios-sdk-v3.1.6/samples/IDCardToolkitObjCApp/IDCardToolkitObjCApp/ViewController/PkiCertificatesViewController.m
//
//  PkiCertificatesViewController.m
//  
//
//  Created by Federal Authority For Identity and Citizenship on 14/11/17.
//  Copyright © 2017 Federal Authority For Identity and Citizenship. All rights reserved.
//

#import "PkiCertificatesViewController.h"

@interface PkiCertificatesViewController ()
@property(nonatomic,strong)Utils * utils;
@end

@implementation PkiCertificatesViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.title=@"Pki Certificates";
    self.navigationItem.hidesBackButton = YES;
    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[[UIImage imageNamed:@"backbutton"] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal]
                                                                             style:UIBarButtonItemStylePlain
                                                                            target:self
                                                                            action:@selector(goBack)];
    self.pkiCertificatesButton.layer.cornerRadius=22.0;
    [self.pkiCertificatesButton.layer setShadowOffset:CGSizeMake(5, 5)];
    [self.pkiCertificatesButton.layer setShadowColor:[[UIColor lightGrayColor] CGColor]];
    [self.pkiCertificatesButton.layer setShadowOpacity:0.5];
   // self.pkiCertificatesButton.layer.borderWidth=1.0;
    
    self.pintext.keyboardType=UIKeyboardTypeNumberPad;
    self.pintext.keyboardAppearance = UIKeyboardAppearanceLight;
    UIToolbar* enterPinNumberToolbar = [[UIToolbar alloc]initWithFrame:CGRectMake(0, 0, 320, 50)];
    enterPinNumberToolbar.barStyle = UIBarStyleBlackTranslucent;
    enterPinNumberToolbar.items = @[[[UIBarButtonItem alloc]initWithTitle:@"Cancel" style:UIBarButtonItemStyleDone target:self action:@selector(enterPinCancelNumberPad)],
                                    [[UIBarButtonItem alloc]initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace target:nil action:nil],
                                    [[UIBarButtonItem alloc]initWithTitle:@"Done" style:UIBarButtonItemStyleDone target:self action:@selector(enterPinDoneWithNumberPad)]];
    [enterPinNumberToolbar sizeToFit];
    self.pintext.inputAccessoryView = enterPinNumberToolbar;
    
    AppDelegate *aDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
    self.model = aDelegate.model;
    self.utils=aDelegate.utils;
    
    // Do any additional setup after loading the view.
}
-(void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:YES];
    
    self.pintext.text=@"";
}
- (IBAction)pkiCertificatesButtonAction:(id)sender {
    
    if ([self.pintext.text length]==0) {
        [AlertView showAlertTitle:ALERT withMessage:ENTER_PIN onView:self];
    }// if
    else {
        if ([self.pintext.text length]<4) {
            [AlertView showAlertTitle:ALERT withMessage:MIN_PIN onView:self];
            self.pintext.text=@"";
        }// if
        else if ([self.pintext.text length]>16) {
            [AlertView showAlertTitle:ALERT withMessage:MAX_PIN onView:self];
            self.pintext.text=@"";
        }// else if
     else {
        [self.utils ShowProgressBar:FETCH_CERTIFICATES andView:self.view];
        [self.pintext setUserInteractionEnabled:YES];
        [self.pintext resignFirstResponder];
            
        NSString *requestId=[Utils generateSecureKey];
        NSString *pinData = self.pintext.text;
        dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^(void){
        @try {
            
            CardReader *cardreader  = [self.model.cardReaderSharedArray objectAtIndex:0];
            DataProtectionKey *data_protection_key=self.model.dataprotectedShared;
            
            NSString *request_handle =  [cardreader prepareRequest:requestId];
            if (request_handle.length==0 || [request_handle isEqualToString:@""] || [request_handle isEqualToString:@"(null)"] ) {
                [AlertView showAlertTitle:ALERT withMessage:REQUEST_HANDLE_EMPTY onView:self];
            }//if
            else {
                NSString *encodePin=[Utils setEncrytion:request_handle data:pinData publickey:[data_protection_key getPublicKey] keylength:[data_protection_key getKeyLength]];
                if (encodePin.length==0 || [encodePin isEqualToString:@""] || [encodePin isEqualToString:@"(null)"] ) {
                    [AlertView showAlertTitle:ALERT withMessage:ENCODE_PIN_EMPTY onView:self];
                }//if
                else {
                    CardCertificates *certificates = [cardreader getPkiCertificates:encodePin];
    
                    dispatch_async(dispatch_get_main_queue(), ^{
                    NSString *xmlString =[certificates getXmlString];
                    NSString *errorMsg =  [self.utils validateToolkitResponse:requestId xmlstring:xmlString];
                    if (![errorMsg isEqualToString:@""] || errorMsg.length>0) {
                        [AlertView showAlertTitle:ALERT withMessage:errorMsg onView:self];
                    }//if
                    else {
                        NSData *authCerData = [[NSData alloc] initWithBytes: [certificates getAuthenticationCertificate] length:[certificates getAuthenticationCertificateLength]];
                        NSString *authStr=[Utils base64forData:authCerData];
                
                        NSData *signCerdata = [[NSData alloc] initWithBytes: [certificates getSigningCertificate] length:[certificates getSigningCertificateLength]];
                        NSString *signStr=[Utils base64forData:signCerdata];
                                
                        self.PinView.hidden=YES;
                        self.authCertificateText.text=[NSString stringWithFormat:@"%@",authStr];
                        self.signCertificateText.text=[NSString stringWithFormat:@"%@",signStr];
                        }//else
                     });// update UI main queue
                   }//else
                 }//else
               }//try
            @catch (NSException *exception) {
                NSString *vgresponse =exception.userInfo[@"ErrorResponse"];
                if (![vgresponse isEqualToString:@""] && vgresponse.length>0) {
                    NSString *errorMsg =  [self.utils validateToolkitResponse:requestId xmlstring:vgresponse];
                    if (![errorMsg isEqualToString:@""] && errorMsg.length>0) {
                        [AlertView showAlertTitle:ALERT withMessage:errorMsg onView:self];
                    }//if
                    else {
                        NSString *exe =[NSString stringWithFormat:@"%@\n%@\n%@",exception.name,exception.description,exception.userInfo];
                        [AlertView showAlertTitle:ALERT withMessage:exe onView:self];
                    }//else
                }//if
                else {
                    NSString *exe =[NSString stringWithFormat:@"%@\n%@\n%@",exception.name,exception.description,exception.userInfo];
                    [AlertView showAlertTitle:ALERT withMessage:exe onView:self];
                }//else
            }// catch
            @finally {
                dispatch_async(dispatch_get_main_queue(), ^{
                    [self.utils DismissProgressBar];
                     self.pintext.text=@"";
                });// update UI main queue
              }// finally
          });//background queue
       }//else
    }//else
}
- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    [self.pintext setUserInteractionEnabled:YES];
    [self.pintext resignFirstResponder];
    return YES;
}
- (void)goBack {
    [self.navigationController popViewControllerAnimated:YES];
}
-(void)enterPinCancelNumberPad {
    [self.pintext resignFirstResponder];
    self.pintext.text = @"";
}
-(void)enterPinDoneWithNumberPad {
    [self.pintext resignFirstResponder];
}
- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
 #pragma mark - Navigation
 
 // In a storyboard-based application, you will often want to do a little preparation before navigation
 - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
 // Get the new view controller using [segue destinationViewController].
 // Pass the selected object to the new view controller.
 }
 */

@end

File: id-card-toolkit-ios-sdk-v3.1.6/id-card-toolkit-ios-sdk-v3.1.6/samples/ICASwiftApp/ICASwiftApp/PkiCertificatesViewController.swift
//
//  PkiCertificatesViewController.swift
//  ICASwiftApp
//
//  Created by Digital Trust Tech on 10/05/22.
//

import UIKit

class PkiCertificatesViewController: UIViewController {
    
    @IBOutlet var PinView: UIView!
    @IBOutlet weak var pintext: UITextField!
    @IBOutlet var pkiCertificatesButton: UIButton!
    @IBOutlet  var authCertificateText: UITextView!
    @IBOutlet  var signCertificateText: UITextView!
    @IBOutlet weak var view_Roundedcorners: UIView!

    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.view_Roundedcorners.cornered_View([.layerMaxXMaxYCorner, .layerMinXMaxYCorner], radius: 20, borderColor: .red, borderWidth: 0)
        
        pkiCertificatesButton.layer.cornerRadius = 22.0
        pkiCertificatesButton.layer.shadowOffset = CGSize(width: 5.0, height: 5.0)
        pkiCertificatesButton.layer.shadowColor = UIColor.lightGray.cgColor
        pkiCertificatesButton.layer.shadowOpacity = 0.5
        
        pintext.keyboardType = .numberPad
        pintext.keyboardAppearance = .light
        let enterPinNumberToolbar = UIToolbar(frame: CGRect(x: CGFloat(0), y: CGFloat(0), width: CGFloat(320), height: CGFloat(50)))
        enterPinNumberToolbar.barStyle = .blackTranslucent
        enterPinNumberToolbar.items = [UIBarButtonItem(title: "Cancel", style: .done, target: self, action: #selector(enterPinCancelNumberPad)), UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil), UIBarButtonItem(title: "Done", style: .done, target: self, action: #selector(enterPinDoneWithNumberPad))]
        enterPinNumberToolbar.sizeToFit()
        pintext.inputAccessoryView=enterPinNumberToolbar
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(true)
        pintext.text=""
        pintext.resignFirstResponder()
    }// viewWillAppear()
        
    @IBAction func pkiCertificatesButtonAction(_ sender: Any) {
        
        if toolkit == nil  {
            self.view.makeToast(Constants.toolkitNotInitilized)
            return
        }//if
        else if cardreader == nil {
            self.view.makeToast(Constants.cardreaderNotConnect)
            return
        }//else if
        
        self.pintext.resignFirstResponder()
        if self.pintext.text?.count==0 {
            AlertView.showAlertTitle(Constants.Alert, withMessage: Constants.EnterPin, onView: self)
        }// if
        else {
            if (self.pintext.text?.count)!<4 {
                AlertView.showAlertTitle(Constants.Alert, withMessage: Constants.MinPin, onView: self)
                self.pintext.text=""
            }// if
            else if (self.pintext.text?.count)!>16 {
                AlertView.showAlertTitle(Constants.Alert, withMessage: Constants.MaxPin, onView: self)
                self.pintext.text=""
            }// else if
            else {
                utils.showProgressBar(Constants.FetchCertificates, andView: self.view)
                let requestId = Utils.generateSecureKey()
                DispatchQueue.global(qos: .background).async {
                    do {
                        let request_handle = try cardreader?.prepareRequest(requestId!)
                        if (request_handle?.isEmpty == true) || (request_handle == "") || (request_handle == "(null)") {
                            AlertView.showAlertTitle(Constants.Alert, withMessage:Constants.Request_Handle_Empty, onView: self)
                        }//if
                        else {
                            let encodePin:String? = Utils.setEncrytion(request_handle, data: self.pintext.text!,publickey: dataProtectionKey.getPublicKey(), keylength: Int32(dataProtectionKey.getKeyLength()))
                            if (encodePin?.isEmpty)! || (encodePin == "") || (encodePin == "(null)") {
                                AlertView.showAlertTitle(Constants.Alert, withMessage:Constants.Encode_Pin_Empty, onView: self)
                            }//if
                            else {
                                certificates = try cardreader?.getPkiCertificates(encodePin!)
                                DispatchQueue.main.async {
                                    let xmlString = certificates.getXmlString()
                                    let errorMsg: String = utils.validateToolkitResponse(requestId, xmlstring: xmlString)
                                    if (errorMsg != "") && (errorMsg.count ) > 0 {
                                        AlertView.showAlertTitle(Constants.Alert, withMessage: errorMsg, onView: self)
                                    }//if
                                    else {
                                        let authCerData = NSData(bytes: certificates.getAuthenticationCertificate(), length: Int(certificates.getAuthenticationCertificateLength()))
                                        let authStr=Utils.base64forData(authCerData as Data)
                                        
                                        let signCerdata = NSData(bytes: certificates.getSigningCertificate(), length: Int(certificates.getSigningCertificateLength()))
                                        let signStr=Utils.base64forData(signCerdata as Data)
                                        
                                        self.PinView.isHidden=true;
                                        self.authCertificateText.text=authStr
                                        self.signCertificateText.text=signStr
                                    }//else
                                }// update UI main queue
                            }//else
                        }//esle
                        DispatchQueue.main.async {
                            utils.dismissProgressBar()
                            self.pintext.text=""
                        }// update UI main queue
                    }//do
                    catch let error as NSError{
                        DispatchQueue.main.async {
                            var vgresponse = String()
                            if error.userInfo["ErrorResponse"] != nil {
                                vgresponse = error.userInfo["ErrorResponse"] as! String
                                if (vgresponse != "") && ((vgresponse.count ) > 0) {
                                    let errorMsg: String = utils.validateToolkitResponse(requestId, xmlstring: vgresponse)
                                    if (errorMsg != "") && ((errorMsg.count ) > 0) {
                                        AlertView.showAlertTitle(Constants.Alert, withMessage:errorMsg, onView: self)
                                    }//if
                                    else {
                                        let err="\(error.domain)\n\(error.code)\n\(error.userInfo)"
                                        AlertView.showAlertTitle(Constants.Alert, withMessage:err,onView: self)
                                    }//else
                                }//if
                                else {
                                    let err="\(error.domain)\n\(error.code)\n\(error.userInfo)"
                                    AlertView.showAlertTitle(Constants.Alert, withMessage:err,onView: self)
                                }//else
                            }//if
                            else {
                                let err="\(error.domain)\n\(error.code)\n\(error.userInfo)"
                                AlertView.showAlertTitle(Constants.Alert, withMessage:err,onView: self)
                            }//else
                            utils.dismissProgressBar()
                            self.pintext.text=""
                        }// update UI main queue
                    }//catch
                }//background queue
            }//else
        }//else
    }
    
    @objc func enterPinCancelNumberPad() {
        pintext.resignFirstResponder()
        pintext.text=""
    }
    
    @objc func enterPinDoneWithNumberPad() {
        pintext.resignFirstResponder()
    }
    
    @IBAction func btn_BackAction(_ sender: Any) {
        navigationController?.popViewController(animated: true)
    }
    
}
